### PR TITLE
Implementation of NPC Dialogues

### PR DESCRIPTION
**Pull request by abimek**
Decided to make a new api instead of using the forms api, more flexible in the future.

### CODE REVIEW

**Review by Flonja — CHANGES_REQUESTED** [review]
A few minor changes

File: server/player/dialogue/button_type.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+// ActivationEnter is the enter activation type.
+func ActivationEnter() ActivationType {
+	return ActivationType{2}
+}
+
+type button uint8
+
+// ButtonType represents the type of Button. URL, COMMAND, and UNKNOWN are all the button types.
```
[the remark this answers is not in the document]
Comment (abimek) [reply server/player/dialogue/button_type.go:31]: URL Button isn't possible in mc as servers can't redicrect a user to a link.

[the remark this answers is not in the document]
Comment (Flonja) [reply server/player/dialogue/button_type.go:31]: Why include it in the documentation then?

File: server/session/player.go [hunk 2]
```diff
@@ -328,6 +329,57 @@ func (s *Session) SendForm(f form.Form) {
...
+func (s *Session) SendDialogue(d dialogue.Dialogue) {
+	h := s.handlers[packet.IDNPCRequest].(*NpcRequestHandler)
+
+	m := d.Menu()
+	h.mu.Lock()
+	h.dialogues[s.c.XUID()] = d
+	h.mu.Unlock()
+
+	s.entityMutex.Lock()
+	RID := s.entityRuntimeIDs[m.NPC()]
```
Comment (T14Raptor) [inline server/session/player.go:343]: Use `(*Session).entityRuntimeID`

File: server/session/player.go [hunk 3]
```diff
@@ -328,6 +329,57 @@ func (s *Session) SendForm(f form.Form) {
...
+}
+
+// CloseDialogue forcefully closes the users current dialogue. The Close method of dialogue.Closer is called when the
+// form closes on the client.
+func (s *Session) CloseDialogue(d dialogue.Dialogue) {
+	h := s.handlers[packet.IDNPCRequest].(*NpcRequestHandler)
+
+	m := d.Menu()
+	s.entityMutex.Lock()
+	RID := s.entityRuntimeIDs[m.NPC()]
```
Comment (T14Raptor) [inline server/session/player.go:364]: Same here

File: server/session/player.go (hunk 2 above)
Comment (abimek) [reply server/session/player.go:343]: you you mean currentEnttiyRunetimeId() ?

**Pull request closed without merging**